File: challenges-in-python/cyclic_sort/missing_number.py
# Solving the "Missing Number" Challenge
# Challenge Description:
# Given an array nums containing n distinct numbers in the range [0, n], 
# return the only number in the range that is missing from the array.

# Approach 1:
# Use the Cyclic Sort pattern to place each number in its correct position, 
# and then iterate through the sorted array to find the missing number.

# Big O:
# O(n) Time | O(1) Space
def find_missing_number_1(nums):
    i, n = 0, len(nums)
    while i < n:
        j = nums[i]
        if j < n and nums[i] != nums[j]:
            nums[i], nums[j] = nums[j], nums[i] # swap
        else:
            i += 1

    # After sorting, the first index where the number is not equal to the index is the missing number
    for i in range(n):
        if nums[i] != i:
            return i
    return n

# Approach 2:
# Use the XOR operation to find the missing number.

# Big O:
# O(n) Time | O(1) Space
def find_missing_number_2(nums):
    n = len(nums)
    x1, x2 = 0, 0
    for i in range(n):
        x1 ^= i # XOR operation
        x2 ^= nums[i] # XOR operation
    return x1 ^ x2 ^ n # XOR operation

# Approach 3:
# Use the Gauss' Formula to find the missing number.

# Big O:
# O(n) Time | O(1) Space
def find_missing_number_3(nums):
    n = len(nums)
    expected_sum = n * (n + 1) // 2
    actual_sum = sum(nums)
    return expected_sum - actual_sum

# if the missing numbers range is [n, m], then the expected sum is:
# expected_sum = ((m - n + 1) * (n + m)) // 2


# Test Cases
def test_find_missing_number():
    assert find_missing_number_1([3, 0, 1]) == 2
    assert find_missing_number_2([0, 1]) == 2
    assert find_missing_number_3([9,6,4,2,3,5,7,0,1]) == 8
    assert find_missing_number_3([0]) == 1

test_find_missing_number()

File: challenges-in-rust/src/cyclic_sort/missing_number.rs
// Solving the "Missing Number" Challenge
// Problem Statement
// Given an array nums containing n distinct numbers in the range [0, n], 
// return the only number in the range that is missing from the array.

// Approach 1:
// We will iterate through the array and place the numbers at their correct index.
// After sorting the array, we will iterate through the array and check if the number is at its correct index.
// If the number is not at its correct index, we will return the number.
// If all the numbers are at their correct index, we will return the length of the array.

// Big O:
// Time Complexity: O(n) | Space Complexity: O(1).
pub fn missing_number(nums: Vec<i32>) -> i32 {
    let mut nums = nums;
    let n = nums.len();
    let mut i = 0;
    while i < n {
        let j = nums[i] as usize;
        if j < n && nums[i] != nums[j] {
            nums.swap(i, j);
        } else {
            i += 1;
        }
    }
    for i in 0..n {
        if nums[i] != i as i32 {
            return i as i32;
        }
    }
    n as i32
}

// Approach 2:
// The problem can also be solved using the sum of the first n natural numbers.
// We will calculate the sum of the first n natural numbers.
// We will iterate through the array and calculate the sum of the given array.
// We will subtract the sum of the given array from the sum of the first n natural numbers.
// The result will be the missing number from the given array.

// Big O:
// Time Complexity: O(n) | Space Complexity: O(1).
pub fn missing_number_2(nums: Vec<i32>) -> i32 {
    let n = nums.len() as i32;
    let mut sum = n * (n + 1) / 2; // sum of the first n natural numbers
    for num in nums {
        sum -= num;
    }
    sum
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_missing_number() {
        assert_eq!(missing_number(vec![3, 0, 1]), 2);
        assert_eq!(missing_number(vec![0, 1]), 2);
    }

    #[test]
    fn test_missing_number_2() {
        assert_eq!(missing_number_2(vec![3, 0, 1]), 2);
        assert_eq!(missing_number_2(vec![0, 1]), 2);
    }
}


